Jenkinsfile

```diff
@@ -283,534 +365,151 @@
-    <run task="partest-noopt">
-      <v n="scala">2.11.8</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fastopt">
-      <v n="scala">2.11.8</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fullopt">
-      <v n="scala">2.11.8</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-noopt">
-      <v n="scala">2.11.11</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fastopt">
-      <v n="scala">2.11.11</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fullopt">
-      <v n="scala">2.11.11</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-noopt">
-      <v n="scala">2.11.12</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fastopt">
-      <v n="scala">2.11.12</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fullopt">
-      <v n="scala">2.11.12</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-noopt">
-      <v n="scala">2.12.0</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fastopt">
-      <v n="scala">2.12.0</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fullopt">
-      <v n="scala">2.12.0</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-noopt">
-      <v n="scala">2.12.1</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fastopt">
-      <v n="scala">2.12.1</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fullopt">
-      <v n="scala">2.12.1</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-noopt">
-      <v n="scala">2.12.2</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fastopt">
-      <v n="scala">2.12.2</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fullopt">
-      <v n="scala">2.12.2</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-noopt">
-      <v n="scala">2.12.3</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fastopt">
-      <v n="scala">2.12.3</v>
-      <v n="java">1.8</v>
-    </run>
-    <run task="partest-fullopt">
-      <v n="scala">2.12.3</v>
-      <v n="java">1.8</v>
-    </run>
-  </matrix>
-
-</ci>
+  '''
+]
+
+def mainJavaVersion = "1.8"
+def otherJavaVersions = [] // should be ["1.6", "1.7"] but that's broken (see #3293)
+def allJavaVersions = otherJavaVersions.clone()
+allJavaVersions << mainJavaVersion
+
+def mainScalaVersion = "2.12.4"
+def mainScalaVersions = ["2.10.2", "2.11.12", "2.12.4", "2.13.0-M2"]
+def otherScalaVersions = [
+  "2.10.3",
+  "2.10.4",
+  "2.10.5",
+  "2.10.6",
+  "2.10.7",
+  "2.11.0",
+  "2.11.1",
+  "2.11.2",
+  "2.11.4",
+  "2.11.5",
+  "2.11.6",
+  "2.11.7",
+  "2.11.8",
+  "2.11.11",
+  "2.11.12",
+  "2.12.0",
+  "2.12.1",
+  "2.12.2",
+  "2.12.3"
+]
+
+// The 'quick' matrix
+def quickMatrix = []
+mainScalaVersions.each { scalaVersion ->
+  allJavaVersions.each { javaVersion ->
+    quickMatrix.add([task: "main", scala: scalaVersion, java: javaVersion])
```

Comment:
Have you considered making function calls out of the tasks rather than a string -> string map? We could use [named arguments](http://docs.groovy-lang.org/latest/html/documentation/#_named_arguments) to pass in the params. I believe this will allow us to reduce duplication going forward (e.g. partest optimization levels, ES versions, etc.)

If you rather want to get something out fast, I'm OK with it, but refactoring this across two branches might be painful going forward.

Reply:
I have considered it, but elected for this variant which deviates less from the previous matrix specification.

I can change it to functions if you prefer.

Reply:
Meh, let's leave it so we can move forward. It's probably worth to schedule a rework of this once the 0.6.x branch is gone anyways.